cmd/restic/integration_test.go

```diff
@@ -359,6 +359,40 @@ func TestBackupNonExistingFile(t *testing.T) {
 	testRunBackup(t, "", dirs, opts, env.gopts)
 }
 
+func TestBackupSelfHealing(t *testing.T) {
+	env, cleanup := withTestEnvironment(t)
+	defer cleanup()
+
+	testRunInit(t, env.gopts)
+
+	p := filepath.Join(env.testdata, "test/test")
+	rtest.OK(t, os.MkdirAll(filepath.Dir(p), 0755))
+	rtest.OK(t, appendRandomData(p, 5))
+
+	opts := BackupOptions{}
+
+	testRunBackup(t, filepath.Dir(env.testdata), []string{filepath.Base(env.testdata)}, opts, env.gopts)
+	testRunCheck(t, env.gopts)
+
+	// remove data file => File with length 110 (= 5 bytes + 32 crypto + 37 bytes pack header entry + 4 bytes header length + 32 bytes crypto for pack header
+	rtest.OK(t, filepath.Walk(env.repo, func(path string, info os.FileInfo, err error) error {
```

Comment:
That looks like a really ugly hack. Please open the repository using `OpenRepository`, then call the backend to list and remove the data pack.

Checking the filesize is another hack. I'd propose to use the treePacks information in the index as a list of packs to keep:
```
	treePacks := restic.NewIDSet()
	for _, idx := range r.idx.All() {
		for _, id := range idx.TreePacks() {
			treePacks.Insert(id)
		}
	}
```

Reply:
It's still a bit hacky as `idx` is private and `restic.Index` does neither expose a `All()` nor a `Treepacks()` method. However, it is much cleaner now.